src/passes/externalContractHandler/externalContractInterfaceInserter.ts

```diff
@@ -0,0 +1,154 @@
+import assert = require('assert');
+import {
+  ContractDefinition,
+  ContractKind,
+  FunctionCall,
+  FunctionDefinition,
+  FunctionKind,
+  getNodeType,
+  Identifier,
+  MemberAccess,
+  SourceUnit,
+  UserDefinedType,
+  UserDefinedTypeName,
+  VariableDeclaration,
+} from 'solc-typed-ast';
+
+import { AST } from '../../ast/ast';
+import { CairoContract } from '../../ast/cairoNodes';
+import { ASTMapper } from '../../ast/mapper';
+import { cloneASTNode } from '../../utils/cloning';
+
+function genContractInterface(
+  contract: ContractDefinition,
+  sourceUnit: SourceUnit,
+  ast: AST,
+): ContractDefinition {
+  const contractId = ast.reserveId();
+  const contractInterface = new CairoContract(
+    contractId,
+    '',
+    `${contract.name}@interface`,
+    sourceUnit.id,
+    ContractKind.Interface,
+    contract.abstract,
+    false,
+    contract.linearizedBaseContracts,
+    contract.usedErrors,
+    new Map(),
+    0,
+  );
+
+  contract.vFunctions.forEach((func) => {
+    // return if func is constructor
+    if (func.kind === FunctionKind.Constructor) return;
+
+    const funcClone = new FunctionDefinition(
+      ast.reserveId(),
+      '',
+      contractInterface.id,
+      func.kind,
+      func.name,
+      func.virtual,
+      func.visibility,
+      func.stateMutability,
+      func.isConstructor,
+      cloneASTNode(func.vParameters, ast),
+      cloneASTNode(func.vReturnParameters, ast),
+      func.vModifiers,
+    );
+    contractInterface.appendChild(funcClone);
+    ast.registerChild(funcClone, contractInterface);
+  });
+  sourceUnit.appendChild(contractInterface);
+  ast.registerChild(contractInterface, sourceUnit);
+  return contractInterface;
+}
+
+function importExternalContract(
+  contract: ContractDefinition,
+  sourceUnit: SourceUnit | undefined,
+  contractInterfaces: Map<number, ContractDefinition>,
+  ast: AST,
+) {
+  const contractSourceUnit = contract.getClosestParentByType(SourceUnit);
+
+  assert(sourceUnit !== undefined, 'Trying to import a definition into an unknown source unit');
+
+  if (contractSourceUnit === undefined || sourceUnit === contractSourceUnit) return;
+
+  if (contract.kind === ContractKind.Library) return;
+  if (contractInterfaces.has(contract.id)) return;
+  contractInterfaces.set(contract.id, genContractInterface(contract, sourceUnit, ast));
+}
+
+export class ExternalContractInterfaceInserter extends ASTMapper {
```

Comment:
I think this pass has the potential to import more interfaces than are necessary. Would it be sufficient to specifically check FunctionCalls whose vExpression is a MemberAccess, and to check the type of the child node of that member access to see if it's a contract?

Reply:
It seems that it is sufficient and all tests are passing, I covered the extra cases to just be safe.